cpp/CMakeLists.txt

```diff
@@ -100,15 +102,28 @@ endif((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") AND(CMAKE_CXX_COMPILER_VERSION VERS
 
 # define flags to enable most warnings and treat them as errors for different compilers
 # add flags to each target separately instead of globally so users have the choice to use their own flags
-if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
-    set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
-        "-Wno-unknown-warning;-Wno-pragmas;-Wall;-Wextra;-Werror;-Wshadow;--pedantic-errors;-Wno-deprecated-copy;-Wno-expansion-to-defined")
-elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
-        "-Wno-unknown-warning-option;-Wall;-Wextra;-Werror;-Wshadow;--pedantic-errors;-Wno-deprecated;-Wno-gnu-zero-variadic-macro-arguments")
-elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
-    set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
-        "/W4;/WX")
+if(MEMILIO_ENABLE_WARNINGS_AS_ERRORS)
+    if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "-Wno-unknown-warning;-Wno-pragmas;-Wall;-Wextra;-Werror;-Wshadow;--pedantic-errors;-Wno-deprecated-copy;-Wno-expansion-to-defined")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "-Wno-unknown-warning-option;-Wall;-Wextra;-Werror;-Wshadow;--pedantic-errors;-Wno-deprecated;-Wno-gnu-zero-variadic-macro-arguments")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "/W4;/WX")
+    endif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+elseif(MEMILIO_ENABLE_WARNINGS)
+    if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "-Wno-unknown-warning;-Wno-pragmas;-Wall;-Wextra;-Wshadow;--pedantic;-Wno-deprecated-copy;-Wno-expansion-to-defined")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "-Wno-unknown-warning-option;-Wall;-Wextra;-Wshadow;--pedantic;-Wno-deprecated;-Wno-gnu-zero-variadic-macro-arguments")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+        set(MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS
+            "/W4")
+    endif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")

```

Comment:
I think this can be simplified by separating the warning flags and the error flags, i.e. if warnings are enabled, add the warning flags, if warnings as errors are enabled, add the error flags. then there is less repetition. It's ok if error flags are enabled and warnings are disabled. I guess --pedantic and --pedantic-errors needs a bit special handling then, but that is the same for clang and gcc and doesn't exist for msvc, that should make things simpler. 
you can use the cmake string or list functions to build up the flags.

something like:
```
if warnings
  add-warning-flags
  if errors
    add-pedantic-errors
  else
    add-pedantic
  endif
endif

if errors
  add-error-flags
endif
```